Adds the pipe sparker

Domic2 opened the pull request:
:cl:
rscadd: Adds pipe sparkers, pipe fires can now be started without heating the mix. Can be controlled via signalers and buttons, or disabled with a screwdriver.
/:cl:

This was somehow easier than I thought it'd be, hopefully I didn't mess anything up too badly.

Unwrenching the sparker will disconnect it from connected buttons.
Albens provided the base sprite, I did the rest.
The sparker ignites the entire pipeline, multiple sparkers are not needed for larger systems.

Domic2 committed:
Adds the pipe sparker

MuckerMayhem commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,131 @@
+	alert_pressure = 350*ONE_ATMOSPHERE
+
+	var/last_spark = 0
+	var/disabled = 0
```
```suggestion
	var/disabled = FALSE
```

MuckerMayhem commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,131 @@
+
+/obj/machinery/atmospherics/pipe/cap/sparker/proc/ignite()
+	playsound(loc, 'sound/machines/click.ogg', 10, 1)
+	if (world.time < last_spark + 50 || !powered() || disabled)
+		return
```
You do this check in `physical_attack_hand()` as well. I suggest just adding a `return` in the first check so that `ignite()` is never called if it fails.

MuckerMayhem replied:
You shouldn't need this check at all now since you are returning if it fails in the proc that calls this.

Domic2 replied:
It's also called by the button trigger (line 33) and process_activation (line 105), I'd probably have to add another proc for the trigger, too.

MuckerMayhem replied:
Might just be good as is then, at least with the return you're not doing the check twice on physical activations.

Domic2 replied:
Indeed!

MuckerMayhem commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,131 @@
+	last_spark = world.time
+
+/obj/machinery/atmospherics/pipe/cap/sparker/physical_attack_hand(mob/user)
+	playsound(src.loc, "button", 30, 1)
```
```suggestion
	playsound(loc, "button", 30, 1)
```

MuckerMayhem commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,131 @@
+	if (world.time < last_spark + 50 || !powered() || disabled)
+		return
+
+	playsound(src.loc, "sparks", 100, 1)
```
```suggestion
	playsound(loc, "sparks", 100, 1)
```

MuckerMayhem reviewed (changes requested):
Noting a lot of inconsistent styling with your `if`'s, do a pass and change all `if(...` -> `if (...`

Domic2 commented:
Should be good now, unless I missed something.

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+/obj/machinery/atmospherics/pipe/cap/sparker/physical_attack_hand(mob/user)
+	playsound(loc, "button", 30, 1)
+	if (world.time < last_spark + 50 || !powered() || disabled)
+		user.visible_message("<span class='notice'>[user] tries to activate the [src], but nothing happens.</span>", "<span class='notice'>You try to activate the [src], but nothing happens.</span>")
```
Use span defines and `\the`

```suggestion
		user.visible_message(
			SPAN_NOTICE("\The [user] tries to activate \the [src], but nothing happens."),
			SPAN_NOTICE("You try to activate \the [src], but nothing happens.")
		)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+		playsound(loc, 'sound/machines/click.ogg', 10, 1)
+		return
+	else
+		user.visible_message("<span class='notice'>[user] activates the [src].</span>", "<span class='notice'>You activate the [src].</span>")
```
```suggestion
		user.visible_message(
			SPAN_NOTICE("\The [user] activates \the [src]."),
			SPAN_NOTICE("You activate \the [src].")
		)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+		add_fingerprint(user)
+		disabled = !disabled
+		if (disabled)
+			user.visible_message("<span class='warning'>[user] has disabled the [src]!</span>", "<span class='warning'>You disable the connection to the [src].</span>")
```
```suggestion
			user.visible_message(
				SPAN_WARNING("\The [user] has disabled \the [src]!"),
				SPAN_WARNING("You disable the connection to \the [src].")
			)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+		if (disabled)
+			user.visible_message("<span class='warning'>[user] has disabled the [src]!</span>", "<span class='warning'>You disable the connection to the [src].</span>")
+		else if (!disabled)
+			user.visible_message("<span class='notice'>[user] has reconnected wiring on the [src]!</span>", "<span class='notice'>You fix the wire connection on the [src].</span>")
```
```suggestion
			user.visible_message(
				SPAN_NOTICE("\The [user] has reconnected wiring on \the [src]!"),
				SPAN_NOTICE("You fix the wire connection on \the [src].")
			)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+
+	if (istype(W, /obj/item/device/assembly/signaler ) && signaler == null)
+		if (disabled)
+			to_chat(user, "<span class='warning'>The [src] is disabled!</span>")
```
```suggestion
			to_chat(user, SPAN_WARNING("\The [src] is disabled!"))
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+			return
+		signaler = W
+		if (signaler.secured)
+			to_chat(user, "<span class='warning'>The [signaler] is secured!</span>")
```
```suggestion
			to_chat(user, SPAN_WARNING("\The [signaler] is secured!"))
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+		signaler.holder = src
+		user.unEquip(signaler)
+		signaler.forceMove(src)
+		user.visible_message("<span class='notice'>[user] connects the [signaler] to the [src].</span>", "<span class='notice'>You connect the [signaler] to the [src].</span>")
```
```suggestion
		user.visible_message(
			SPAN_NOTICE("\The [user] connects \the [signaler] to \the [src]."),
			SPAN_NOTICE("You connect \the [signaler] to \the [src].")
		)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+	if (isScrewdriver(W) && signaler)
+		signaler.holder = null
+		signaler.dropInto(loc)
+		user.visible_message("<span class='warning'>[user] disconnects the [signaler] from the [src].</span>", "<span class='warning'>You disconnect the [signaler] from the [src].</span>")
```
```suggestion
		user.visible_message(
			SPAN_WARNING("\The [user] disconnects \the [signaler] from \the [src]."),
			SPAN_WARNING("You disconnect \the [signaler] from \the [src].")
		)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,133 @@
+		update_icon()
+		return
+	if (isWrench(W) && (signaler || disabled))
+		to_chat(user, "<span class='notice'>Remove signalers and check the wiring before unwrenching the [src].</span>")
```
```suggestion
		to_chat(user, SPAN_NOTICE("Remove signalers and check the wiring before unwrenching \the [src]."))
```

Domic2 commented:
Alright, looks good, let's hope there isn't anything else I didn't know about

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,151 @@
+	playsound(loc, "sparks", 100, 1)
+	use_power_oneoff(2000)
+	flick ("pipe-sparker-spark", src)
+	parent.air.react(null, 1, 1)//full bypass
```
Use `TRUE`/`FALSE` instead of `1`/`0`

```suggestion
	parent.air.react(null, TRUE, TRUE)//full bypass
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,151 @@
+
+	playsound(loc, "sparks", 100, 1)
+	use_power_oneoff(2000)
+	flick ("pipe-sparker-spark", src)
```
Code style

```suggestion
	flick("pipe-sparker-spark", src)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,151 @@
+
+/obj/machinery/atmospherics/pipe/cap/sparker/proc/ignite()
+	playsound(loc, 'sound/machines/click.ogg', 10, 1)
+	if (world.time < last_spark + 50 || !powered() || disabled)
+		return
```
This check is repeated in a number of places - I'd suggest adding a `can_ignite()` proc that just runs this check and then calling that.

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,151 @@
+
+/obj/machinery/atmospherics/pipe/cap/sparker/physical_attack_hand(mob/user)
+	playsound(loc, "button", 30, 1)
+	if (world.time < last_spark + 50 || !powered() || disabled)
```
See above about the `can_ignite()` suggestion

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,151 @@
+		)
+		playsound(loc, 'sound/machines/click.ogg', 10, 1)
+		return
+	else
+		user.visible_message(
+			SPAN_NOTICE("\The [user] activates \the [src]."),
+			SPAN_NOTICE("You activate \the [src].")
+		)
```
The `return` renders the `else` redundant.

```suggestion
	user.visible_message(
		SPAN_NOTICE("\The [user] activates \the [src]."),
		SPAN_NOTICE("You activate \the [src].")
	)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,151 @@
+			to_chat(user, SPAN_WARNING("\The [signaler] is secured!"))
+			signaler = null
+			return
+		signaler.holder = src
```
I'm not sure this is valid or intended use. The type of `src` does not match the type defined for the `holder` var.

SierraKomodo reviewed (changes requested):
Also needs squashing.

Domic2 commented:
I can't believe this worked, wow.
Well, it's squashed, added the "mholder" var so signalers can now properly interact with machinery, and all other requested changes are done.
I replaced a few "\the" with "the" because they weren't adding "the" where I wanted them to add one, hopefully that's alright.

Domic2 commented:
\The wasn't adding a 'the' because the name was wrong, works properly now.

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,156 @@
+	icon_state = "pipe-sparker"
+	alpha = 128
+
+/obj/machinery/atmospherics/pipe/cap/sparker/proc/can_ignite()
+	if ((world.time < last_spark + 50) || !powered() || disabled)
+		return 1
+	return 0
```
```suggestion
		return TRUE
	return FALSE
```

SierraKomodo replied:
The proc is named `can_ignite()` but the way it's used seems to indicate it's checking if you can _not_ ignite. I'd suggest either renaming the proc or changing the return values to fit the current name.

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,156 @@
+			return
+		add_underlay(T, node, dir)
+
+/obj/machinery/atmospherics/pipe/cap/sparker/set_color(var/new_color)
```
```suggestion
/obj/machinery/atmospherics/pipe/cap/sparker/set_color(new_color)
```

SierraKomodo commented on code/modules/atmospherics/components/pipesparker.dm:
```diff
@@ -0,0 +1,156 @@
+/obj/machinery/atmospherics/pipe/cap/sparker/set_color(var/new_color)
+	return
+
+/obj/machinery/atmospherics/pipe/cap/sparker/color_cache_name(var/obj/machinery/atmospherics/node)//returns to the original
```
```suggestion
/obj/machinery/atmospherics/pipe/cap/sparker/color_cache_name(obj/machinery/atmospherics/node)//returns to the original
```

Domic2 commented:
All done, not sure what the Maps(away_sites_testing) check found wrong, though.
